### PR TITLE
Show the correct build logs for specific results

### LINKED ISSUE #3849: Load old build logs does not work
### Describe the bug

Loading old build failed results (build logs) depends on the participation id, therefore it is not possible to load build logs for an old submission.

### To Reproduce

1. Participate in a programming exercise multiple times with one build failed and one non build failed submission/result
2. Try to open the build failed result in the history (student or instructor view)
3. Notice that it won't work with an exception

### Expected behavior

It should also work for old results.
The REST call needs to be changed to depend on the result id (or submission id) and not on the participation id.

### Screenshots

_No response_

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by MaisiKoleni**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage) 
- [x] Client: I documented the TypeScript code using JSDoc style. 
- [x] Client: I added multiple screenshots/screencasts of my UI changes 
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context
This resolves #3849. The problem here is, that build logs are stores based on the latest build for each individual submission, but the build logs are only available on participation level. This results in older failed submissions showing the latest submission's build logs (which might not exist).

### Description
- Added a `resultId` as an optional parameter, as discussed in the issue.
- Added integration tests for result-specific build logs.
- Make all client and server tests pass with the changes.

### Steps for Testing
- Make sure the issue cannot be reproduced anymore, you can also be creative here and check what happens with multiple failed submissions. Just be aware that always the latest result of each submission counts.
- You still get the build logs in the online code editor when working on a programming exercise, and it fails (e.g. due to compilation errors)

### Review Progress

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

### Test Coverage
- Server almost 100%:
  image
- Client service now at 100%
  image

### Screenshots
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java [hunk 1]
```diff
@@ -341,28 +344,38 @@ boolean canAccessRepository(Long participationId) throws IllegalArgumentExceptio
...
+        ProgrammingSubmission programmingSubmission = (ProgrammingSubmission) participation.getSubmissions().stream().findFirst().orElse(null);
+        // If a resultId is specified and the ID does not belong to the latest result, find the corresponding submission. Otherwise use the latest submission.
+        if (resultId.isPresent() && (programmingSubmission == null || programmingSubmission.getResults().stream().noneMatch(r -> resultId.get().equals(r.getId())))) {
+            // Note: if the submission was null, this will fail with a good message.
+            programmingSubmission = programmingSubmissionRepository.findByResultIdElseThrow(resultId.get());
+            if (!Objects.equals(participation.getId(), programmingSubmission.getParticipation().getId())) {
+                // The result of the given ID must belong to the participation
+                log.warn("Participation ID {} tried to access the build logs of another participation's submission with ID {}.", participation.getId(),
+                        programmingSubmission.getId());
+                return forbidden();
```
Comment (MDK5) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java:356]: I'd think bad request would be more fitting, given the numbers of the request dont match, its not really that a permission was checked

**Review by julian-christl — APPROVED** [review]
Tested on TS1, the issue can't get reproduced anymore and it still works when solving the exercise online.

**Review by MDK5 — APPROVED** [review]
Reviewed code in testing session, changes look good to me, only one detail:

**Review by Malyuk-A — APPROVED** [review]
Tested on TS1 during a testing session. Works as described 👍

**Review by ndricimrr — APPROVED** [review]
Tested on TS1. Can properly view the build logs for the multiple results.  No issues found

**Review by rs-maker — APPROVED** [review]
Code looks good :+1: .

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java (hunk 1 above)
Comment (jpbernius) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java:356]: +1

**Review by jpbernius — CHANGES_REQUESTED** [review]

**Review by julian-christl — APPROVED** [review]
Reapprove.
Code changes are too trivial for manual tests and look good.

**Review by FrankeLukas — APPROVED** [review]
Code changes lgtm.

**Pull request merged**